### PR TITLE
Implement parent child relationships in `Data_Generation` 

### PR DESCRIPTION
**Pull request by brynnz22**
For issue https://github.com/microbiomedata/nmdc-schema/issues/1144. Allows DataGeneration to have multiple steps.

### CODE REVIEW

File: src/schema/nmdc.yaml [hunk 1]
```diff
@@ -2643,6 +2644,16 @@ slots:
   vendor:
     range: InstrumentVendorEnum
 
+  has_data_generation_parts:
```
Comment (mslarae13) [inline src/schema/nmdc.yaml:2647]: use has_process_parts

has_process_parts will be updated to not have a domain 
has_process_parts slot will be not required, ProtocolExecution will update its slot usage to say has_process_part is required

**Review by mslarae13 — COMMENTED** [review]
@anastasiyaprymolenna please reference this in your update PR too

**Comment by brynnz22** [comment]
Added `process_parts` to the `DataGeneration` class. @anastasiyaprymolenna made `has_process_parts` optional, but required for the `ProtocolExecution` class. This is ready for review.

**Comment by brynnz22** [comment]
Now changing to use `part_of` for child `DataGeneration` to have `range` `DataGeneration`, and change the current `part_of` slot that has `range` `Study` to be `was_informed_by`.

**Comment by brynnz22** [comment]
@brynnz22 change `was_informed_by` to `study_source`

**Comment by brynnz22** [comment]
@aclum @cmungall I change the `was_informed_by` to be a new slot called `associated_study`. I made this the name since we have a similar slot called `associated_dois` that function in a similar way. If you approve this, I think we will need to make an issue to update all of the places that currently use `part_of` to link to a study to use `associated_study` instead. The only place I can think of that this will not be applicable to is linking a child study to a parent study, in which case, I believe `part_of` is the right choice. This is ready for review! Thanks.

**Comment by brynnz22** [comment]
All in all, since this PR involved a lot of iteration: `part_of` slot in the `DataGeneration` class was changed to from having a value of `Study` to having a `DataGeneration` value to link parts of a data generation (e.g. included sequential sequencing steps) to a parent `DataGeneration` and a new slot was added, called `associated_study` to link `DataGeneration` to its corresponding study. The changes to `process_parts` were in an iteration of this change where we thought we might use `process_parts` to model the sequenctial sequencing, but ultimately decided against it. @anastasiyaprymolenna  made this globally optional, and only required as part of the `slot_usage` for `ProtocolExecution`.

File: src/data/valid/DataGeneration-part_of.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+id: nmdc:datgen-99-zUCd5N
```
Comment (aclum) [inline src/data/valid/DataGeneration-part_of.yaml:1]: typecode should be dgns not datgen. I'm surprised this validates @turbomam

File: src/schema/nmdc.yaml [hunk 3]
```diff
@@ -1869,11 +1872,13 @@ classes:
         comments:
           - pattern should allow typecode for Biosample and ProcessedSample
       part_of:
-        range: Study
-        pattern: "^nmdc:sty-[0-9][a-z]{0,6}[0-9]-[A-Za-z0-9]{1,}(\\.[A-Za-z0-9]{1,})*(_[A-Za-z0-9_\\.-]+)?$" # better applied as a structured_pattern, but even that might get out of sync from the asserted Study structured_pattern
+        range: DataGeneration
+        pattern: "^nmdc:datgen-[0-9][a-z]{0,6}[0-9]-[A-Za-z0-9]{1,}(\\.[A-Za-z0-9]{1,})*(_[A-Za-z0-9_\\.-]+)?$" # better applied as a structured_pattern, but even that might get out of sync from the asserted Study structured_pattern
```
Comment (aclum) [inline src/schema/nmdc.yaml:1897]: I think we should remove this in favor of enforcing the pattern in the Classes and then specifying the range as you have done.

File: nmdc_schema/nmdc_materialized_patterns.schema.json [hunk 4]
```diff
@@ -5909,11 +6067,8 @@
                 },
                 "part_of": {
                     "description": "Links a resource to another resource that either logically or physically includes it.",
-                    "items": {
-                        "pattern": "^nmdc:sty-[0-9][a-z]{0,6}[0-9]-[A-Za-z0-9]{1,}(\\.[A-Za-z0-9]{1,})*(_[A-Za-z0-9_\\.-]+)?$",
-                        "type": "string"
-                    },
-                    "type": "array"
+                    "pattern": "^nmdc:datgen-[0-9][a-z]{0,6}[0-9]-[A-Za-z0-9]{1,}(\\.[A-Za-z0-9]{1,})*(_[A-Za-z0-9_\\.-]+)?$",
```
Comment (aclum) [inline nmdc_schema/nmdc_materialized_patterns.schema.json:6070]: can pattern be taken out in favor of enforcing this via range where the value is a class which already has a pattern defined.

File: nmdc_schema/nmdc_materialized_patterns.schema.json [hunk 5]
```diff
@@ -7367,6 +7522,11 @@
                     "$ref": "#/$defs/AnalyteCategoryEnum",
                     "description": "The type of analyte(s) that were measured in the data generation process and analyzed\n  in the Workflow Chain\n"
                 },
+                "associated_study": {
+                    "description": "The study associated with a resource.",
+                    "pattern": "^nmdc:sty-[0-9][a-z]{0,6}[0-9]-[A-Za-z0-9]{1,}(\\.[A-Za-z0-9]{1,})*(_[A-Za-z0-9_\\.-]+)?$",
```
Comment (aclum) [inline nmdc_schema/nmdc_materialized_patterns.schema.json:7527]: same comment as above wrt not having a pattern, instead enforce via a range

File: nmdc_schema/nmdc_materialized_patterns.schema.json [hunk 6]
```diff
@@ -7449,11 +7609,8 @@
                 },
                 "part_of": {
                     "description": "Links a resource to another resource that either logically or physically includes it.",
-                    "items": {
-                        "pattern": "^nmdc:sty-[0-9][a-z]{0,6}[0-9]-[A-Za-z0-9]{1,}(\\.[A-Za-z0-9]{1,})*(_[A-Za-z0-9_\\.-]+)?$",
-                        "type": "string"
-                    },
-                    "type": "array"
+                    "pattern": "^nmdc:datgen-[0-9][a-z]{0,6}[0-9]-[A-Za-z0-9]{1,}(\\.[A-Za-z0-9]{1,})*(_[A-Za-z0-9_\\.-]+)?$",
```
Comment (aclum) [inline nmdc_schema/nmdc_materialized_patterns.schema.json:7612]: We decided not to have a typecode for datgen, instead there are typecodes dgnt and dtms for the subclasses of data generation

**Review by aclum — CHANGES_REQUESTED** [review]

**Comment by brynnz22** [comment]
Going to redo this PR due to crazy merge conflicts. Will retag everyone.

**Pull request closed without merging**